File: sdk/javascript/src/flags/variant-contract.ts
/** Strict bounds shared by every APDL weighted-variant runtime. */

export const MAX_VARIANTS = 10;
export const MAX_VARIANT_WEIGHT = 9_007_199_254_740_991;
export const MAX_TOTAL_VARIANT_WEIGHT = 9_007_199_254_740_991;

interface WeightedValue {
  weight?: unknown;
}

export function hasCanonicalVariantWeights(
  variants: readonly WeightedValue[]
): boolean {
  if (variants.length === 0 || variants.length > MAX_VARIANTS) {
    return false;
  }

  let totalWeight = 0;
  for (const variant of variants) {
    const weight = variant.weight;
    if (
      typeof weight !== 'number'
      || !Number.isSafeInteger(weight)
      || weight < 0
      || weight > MAX_VARIANT_WEIGHT
      || totalWeight > MAX_TOTAL_VARIANT_WEIGHT - weight
    ) {
      return false;
    }
    totalWeight += weight;
  }

  return totalWeight > 0;
}

File: sdk/python/apdl/flags/variant_contract.py
"""Strict bounds shared by every APDL weighted-variant runtime."""

from __future__ import annotations

from collections.abc import Sequence


MAX_VARIANTS = 10
MAX_VARIANT_WEIGHT = 9_007_199_254_740_991
MAX_TOTAL_VARIANT_WEIGHT = 9_007_199_254_740_991


def validate_variant_weight_contract(weights: Sequence[object]) -> None:
    """Reject weights that JavaScript cannot represent or sum exactly."""
    if not weights:
        raise ValueError("variants must contain at least one variant")
    if len(weights) > MAX_VARIANTS:
        raise ValueError(f"variants must contain at most {MAX_VARIANTS} entries")

    total_weight = 0
    for weight in weights:
        if (
            type(weight) is not int
            or weight < 0
            or weight > MAX_VARIANT_WEIGHT
        ):
            raise ValueError(
                "variant weight must be a nonnegative JavaScript-safe integer"
            )
        total_weight += weight
        if total_weight > MAX_TOTAL_VARIANT_WEIGHT:
            raise ValueError(
                "total variant weight exceeds the JavaScript-safe integer limit"
            )

    if total_weight <= 0:
        raise ValueError(
            "variant weights must contain at least one positive weight"
        )
